Developer workflow (#653)
iota-community/iota-wiki: internal/learn/contribute-to-wiki/getting_started/developer.md

# Writing as a Dev

> As a developer, you are responsible for documentation of your project. You are the one to choose its content, write it, and keep it updated with every release of your software. The [wiki maintainers](https://wiki.iota.org/team) and technical writers from IOTA Foundation will help you to coordinate the work, setup your project, and bring it to the standard.

Reviewer: ```suggestion
As a developer, you are responsible for the documentation of your project. You are the one to choose its content, write it, and keep it updated with every release of your software. The [wiki maintainers](https://wiki.iota.org/team) and technical writers from IOTA Foundation will help you to coordinate the work, set up your project, and bring it to standard.
```

Reviewer: or your project's documentation

## Make a Friend

> Before you start, contact a technical writer. We could then take a look at your project, assess its state, give you an advice, and plan the work together. Reach out to us through the `#iota-wiki` Discord channel,the `#dev-documentation` Slack channel, Slack DMs or via e-mail — whatever you have access to.

Reviewer: ```suggestion
Before you start, contact a technical writer. We could then take a look at your project, assess its state, give you advice, and plan the work together. Reach out to us through the `#iota-wiki` Discord channel, the `#dev-documentation` Slack channel, Slack DMs, or via e-mail.
```
"— whatever you have access to " doesn't seem to add much to the sentence and ends with a preposition.

> Active technical writers:

Reviewer: I don't know about yourself, but please add my email here in case any community member wants to contact me. 
  lucas.tortora@iota.org

See [Setting Up Documentation](../how_tos/documentation_setup.md).

## Write Documentation

There are [four kinds of documentation](../reference/structure/overview.md). The reference can be generated automatically from inline comments in your code: say, with [rustdoc](https://doc.rust-lang.org/rustdoc/index.html), [typedoc](https://typedoc.org/), [javadoc](https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javadoc.html), or [pydoc](https://docs.python.org/3/library/pydoc.html). Tutorials are handled by the IOTA Foundation's DevX team. This leaves you with [how to guides](../how_tos/how_to_write_how_tos.md) and explanation articles.

> You do not have to write perfectly. It is a technical writer's job to produce a perfectly polished article. But we need to base our writing on something, and when it comes to a brand new feature or a code change, you momentarily become the world's only expert in it. If you write a rough, flawed article when release your software, your users would already prefer it over a blank page. And it is much less work for technical writers to streamline your text a bit instead of doing our own research of your project.

Reviewer 2: ```suggestion
You do not have to write perfectly. It is a technical writer's job to produce a perfectly polished article. But we need to base our writing on something, and when it comes to a brand new feature or a code change, you momentarily become the world's only expert in it. Keep in mind your users prefer even a rough, flawed article over a blank page when you release your software. And it is much less work for technical writers to streamline your text a bit instead of doing our own research of your project.
```

> You do not have to write perfectly. It is a technical writer's job to produce a perfectly polished article. But we need to base our writing on something, and when it comes to a brand new feature or a code change, you momentarily become the world's only expert in it. If you write a rough, flawed article to the release of your software, your users would already prefer it over a blank page. And it is much less work for technical writers to streamline your text a bit instead of doing our own research of your project.

Reviewer: ```suggestion
You do not have to write perfectly. It is a technical writer's job to produce a perfectly polished article. But we need to base our writing on something, and when it comes to a brand new feature or a code change, you momentarily become the world's only expert in it. If you write a rough, flawed article to release your software, your users would already prefer it over a blank page. And it is much less work for technical writers to streamline your text a bit instead of doing our own research of your project.
```

Reviewer: or
```suggestion
You do not have to write perfectly. It is a technical writer's job to produce a perfectly polished article. But we need to base our writing on something, and when it comes to a brand new feature or a code change, you momentarily become the world's only expert in it. If you write a rough, flawed article when you release of your software, your users would already prefer it over a blank page. And it is much less work for technical writers to streamline your text a bit instead of doing our own research of your project.
```

> ### Write Documentation As a Routine 

Reviewer: What do you think of  "Make a Habit of Writing" or something? Routine sounds a bit tedious to some.

> Documentation is not optional. If a feature is new and undocumented, then nobody knows that they could use it. This feature doesn't exist. If you changed an existing feature and forgot to update the documentation, then users will follow your instructions just to fail. From their point of view, you have introduced a bug. Documentation is fused with your code, and you should treat it as inherent part of your job, in the same right as coding and testing is.

Reviewer: ```suggestion
Documentation is not optional. If a feature is new and undocumented, nobody knows they can use it. This feature does not exist. If you change an existing feature and forget to update the documentation, users will follow your instructions and fail. From their point of view, you have introduced a bug. Documentation is fused with your code, and you should treat it as an inherent part of your job, in the same way as coding and testing.
```

> The natural way to plan documentation is to create a ticket before every release, to treat it as a separate product feature which must be written in a very odd programming language. We would instead suggest your team to treat documentation as a mandatory part of every task:

Reviewer: ```suggestion
The natural way to plan documentation is to create a ticket before every release, to treat it as a separate product feature that you must write in a very odd programming language. We would instead suggest your team treats documentation as a mandatory part of every task:
```

|Documentation as a part of every ticket           |Documentation as a separate ticket                                                |
|---                                               |---                                                                               |
|You have it listed for every feature by default.  | You can forget to create it.                                                     |
|You release the feature documented.               | You can keep the ticket at the bottom of the backlog forever and never write it. |

> |The polished version is in the work right away.   | Documentation tickets get deprioritized over features and bugfixes. TWs get involved weeks later, shortly before a major version release.  |

Reviewer: ```suggestion
|The polished version is in the works right away.   | Documentation tickets get deprioritized over features and bug fixes. TWs get involved weeks later, shortly before a major version release.  |
```

|You write it while it is fresh.                   | You will have to recall every detail later.                                      |
|You write one small piece at a time.              | You write in huge intimidating chunks.                                           |                                     

> |You write regularily and build a habit.           | You write once in a few months and you treat writing as an unpleasant odd job.   |

Reviewer: ```suggestion
|You write regularly and build a habit.           | You write once every few months and treat writing as an unpleasant odd job.   |
```

## Further Reads

Once again, technical writers are here to help. To let us know that you got some work for us, see [Request a Technical Writer](../how_tos/request_tw.md).

Technical writers follow the [full style guide](../reference/style/manifesto.md), but we have prepared a [single-page version](./quick_advice.md) of it for you.

> We also keep a list of techniques that might help you: see [Writing Tools](../explanations/tools/).

Reviewer: ```suggestion
We also keep a list of  [writing tools](../explanations/tools/) that might help you.
```

See [Setting Up Documentation](../how_tos/documentation_setup.md).

## Write Documentation

> There are [four kinds of documentation](../reference/structure/overview.md). The reference can be generated automatically from inline comments in your code: say, with [rustdoc](https://doc.rust-lang.org/rustdoc/index.html), [typedoc](https://typedoc.org/), [javadoc](https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javadoc.html), or [pydoc](https://docs.python.org/3/library/pydoc.html). Tutorials are handled by the IOTA Foundation's DevX team. This leaves you with [how to guides](../how_tos/how_to_write_how_tos.md) and explanation articles.

Reviewer: ```suggestion
There are [four kinds of documentation](../reference/structure/overview.md). You can generate [reference](../reference/structure/reference) automatically from inline comments in your code using: 

* [rustdoc](https://doc.rust-lang.org/rustdoc/index.html)
* [typedoc](https://typedoc.org/)
* [javadoc](https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javadoc.html)
* [pydoc](https://docs.python.org/3/library/pydoc.html). 

[Tutorials](../reference/structure/tutorials) are handled by the IOTA Foundation's DevX team. This leaves you with [how to guides](../how_tos/how_to_write_how_tos.md) and [explanation articles](../reference/structure/explanations).
```

Author: I wanted to have something before we have a dedicated article, but I would prefer to keep it as a paragraph for now to keep the flow. We'll replace it with a link to the article afterwards anyway.
